### PR TITLE
Python: Bump to Pydantic v2

### PR DESCRIPTION
**Pull request by Fokko**

### CODE REVIEW

File: python/pyiceberg/transforms.py [hunk 1]
```diff
@@ -105,17 +104,17 @@ def _transform_literal(func: Callable[[L], L], lit: Literal[L]) -> Literal[L]:
     return literal(func(lit.value))
 
 
-class Transform(IcebergBaseModel, ABC, Generic[S, T]):
+class Transform(RootModel, ABC):
```
Comment (rdblue) [inline python/pyiceberg/transforms.py:107]: Why does this lose the generics?

File: python/pyiceberg/types.py [hunk 2]
```diff
@@ -34,27 +34,33 @@
...
 from pyiceberg.utils.parsing import ParseNumberFromBrackets
-from pyiceberg.utils.singleton import Singleton
 
 DECIMAL_REGEX = re.compile(r"decimal\((\d+),\s*(\d+)\)")
 FIXED = "fixed"
 FIXED_PARSER = ParseNumberFromBrackets(FIXED)
 
 
-class IcebergType(IcebergBaseModel, Singleton):
+class IcebergType(IcebergBaseModel):
```
Comment (rdblue) [inline python/pyiceberg/types.py:63]: Is this a change we should apply anyway since not all `IcebergType` subclasses are `Singleton`?

File: python/pyiceberg/table/snapshots.py [hunk 3]
```diff
@@ -128,5 +119,5 @@ class MetadataLogEntry(IcebergBaseModel):
 
 
 class SnapshotLogEntry(IcebergBaseModel):
-    snapshot_id: str = Field(alias="snapshot-id")
+    snapshot_id: int = Field(alias="snapshot-id")
```
Comment (rdblue) [inline python/pyiceberg/table/snapshots.py:122]: Looks like this was incorrect? Should we fix this change in main?

File: python/pyiceberg/types.py [hunk 4]
```diff
@@ -232,31 +238,40 @@ class NestedField(IcebergType):
...
     doc: Optional[str] = Field(default=None, repr=False)
-    initial_default: Any = Field(alias="initial-default", repr=False)
+    initial_default: Optional[Any] = Field(alias="initial-default", default=None, repr=False)
 
     def __init__(
         self,
         field_id: Optional[int] = None,
         name: Optional[str] = None,
-        field_type: Optional[IcebergType] = None,
+        field_type: Optional[Type[IcebergType]] = None,
```
Comment (rdblue) [inline python/pyiceberg/types.py:250]: What is happening here? Isn't the type an instance of `IcebergType`?

File: python/pyiceberg/types.py (hunk 2 above)
Comment (Fokko) [reply python/pyiceberg/types.py:63]: Currently, this is the case. The singleton will hash the arguments and if there is already an instance with the same arguments, it will be returned. I'm happy to change this to exclude:

- NestedField
- StructType
- MapType
- ListType

Also, the FixedType, DecimalType, etc could be excluded.

Comment (Fokko) [reply python/pyiceberg/types.py:63]: (I've removed this singleton temporarily to make sure that it didn't interfere with Pydantic 2.0)

File: python/pyiceberg/types.py (hunk 4 above)
Comment (Fokko) [reply python/pyiceberg/types.py:250]: This is incorrect indeed,

File: python/pyiceberg/types.py [hunk 5]
```diff
@@ -232,31 +238,40 @@ class NestedField(IcebergType):
 
     field_id: int = Field(alias="id")
     name: str = Field()
-    field_type: IcebergType = Field(alias="type")
+    field_type: IcebergSubTypes = Field(alias="type")
```
Comment (Fokko) [inline python/pyiceberg/types.py:241]: This has been fixed in https://github.com/pydantic/pydantic/issues/6403

File: python/pyiceberg/table/metadata.py [hunk 6]
```diff
@@ -42,55 +45,79 @@
...
+def cleanup_snapshot_id(data: Dict[str, Any]) -> Dict[str, Any]:
+    """Runs before validation."""
+    if CURRENT_SNAPSHOT_ID in data and data[CURRENT_SNAPSHOT_ID] == -1:
+        # We treat -1 and None the same, by cleaning this up
+        # in a pre-validator, we can simplify the logic later on
+        data[CURRENT_SNAPSHOT_ID] = None
+    return data
+
+
+def check_schemas(table_metadata: TableMetadata) -> TableMetadata:
```
Comment (Fokko) [inline python/pyiceberg/table/metadata.py:76]: Changed a lot of these validators to `after`, instead of `before`. The `before` would validate on the Python dictionaries without any alias'es applied. The `after` uses the metadata object itself, which is more safe and less error prone.

Comment (Fokko) [reply python/pyiceberg/table/metadata.py:76]: This is also because the validators are refactored quite a bit.

Comment (Fokko) [reply python/pyiceberg/table/metadata.py:83]: The `before` accept an dict with the unprocessed entries, the `after` an actual metadata object. I've switched the ordering because I think it makes sense to first have the `before`'s, and then the `after`'s

File: python/pyiceberg/table/snapshots.py [hunk 7]
```diff
@@ -59,34 +58,19 @@ class Summary(IcebergBaseModel):
     like snapshot expiration, to skip processing certain snapshots.
     """
 
-    __root__: Dict[str, Union[str, Operation]]
+    operation: Operation = Field()
     _additional_properties: Dict[str, str] = PrivateAttr()
 
-    @root_validator
```
Comment (Fokko) [inline python/pyiceberg/table/snapshots.py:65]: A lot of changes here because the additional fields in the operation `dict` are considered properties. Since the `@root_validator` is not there anymore, we completely had to rewrite this. For the better I'd say.

File: python/pyiceberg/transforms.py [hunk 8]
```diff
@@ -106,47 +100,37 @@ def _transform_literal(func: Callable[[L], L], lit: Literal[L]) -> Literal[L]:
...
-    @classmethod
-    def __get_validators__(cls) -> Generator[AnyCallable, None, None]:
-        """Called to validate the input of the Transform class."""
-        # one or more validators may be yielded which will be called in the
-        # order to validate the input, each validator will receive as an input
-        # the value returned from the previous validator
-        yield cls.validate
-
-    @classmethod
-    def validate(cls, v: Any) -> IcebergBaseModel:
```
Comment (Fokko) [inline python/pyiceberg/transforms.py:127]: This changed in 2.0: https://docs.pydantic.dev/latest/migration/#validate_arguments-has-been-renamed-to-validate_call

File: python/pyiceberg/types.py [hunk 9]
```diff
@@ -332,33 +382,46 @@ class ListType(IcebergType):
...
-        data["element_required"] = data["element-required"] if "element-required" in data else element_required
-        data["element_field"] = NestedField(
+        data["element-id"] = data["element-id"] if "element-id" in data else element_id
+        data["element"] = element or data["element_type"]
+        data["element-required"] = data["element-required"] if "element-required" in data else element_required
+        super().__init__(**data)
+        self._hash = hash(data.values())
+
+    @cached_property
+    def element_field(self) -> NestedField:
```
Comment (Fokko) [inline python/pyiceberg/types.py:402]: I think this is much nicer than pre-creating a private attribute.

File: python/tests/table/test_metadata.py [hunk 10]
```diff
@@ -98,19 +97,6 @@ def test_from_dict_v2_parse_raw(example_table_metadata_v2: Dict[str, Any]) -> No
...
 
 
-@pytest.mark.parametrize(
-    "table_metadata, expected_version",
-    [
-        (EXAMPLE_TABLE_METADATA_V1, 1),
-        (EXAMPLE_TABLE_METADATA_V2, 2),
-    ],
-)
-def test_table_metadata_factory(table_metadata: Dict[str, Any], expected_version: int) -> None:
```
Comment (Fokko) [inline python/tests/table/test_metadata.py:108]: The factory was not needed

**Comment by Fokko** [comment]
Did some local tests, and it looks great:
image

**Comment by Fokko** [comment]
I also did some benchmarking, and it is faster:

This branch:
```
➜  python git:(fd-bump-pydantic) ✗ hyperfine --warmup 1 "python3 tests/test_benchmark.py" 
Benchmark 1: python3 tests/test_benchmark.py
  Time (mean ± σ):      1.462 s ±  0.006 s    [User: 1.433 s, System: 0.026 s]
  Range (min … max):    1.454 s …  1.472 s    10 runs
```

Current master:
```
➜  python git:(master) ✗ hyperfine --warmup 1 "python3 tests/test_benchmark.py"
Benchmark 1: python3 tests/test_benchmark.py
  Time (mean ± σ):      1.961 s ±  0.011 s    [User: 1.933 s, System: 0.024 s]
  Range (min … max):    1.944 s …  1.978 s    10 runs
```

This parses 10k json strings:
```python
import json

from pyiceberg.table.metadata import TableMetadataUtil

EXAMPLE_TABLE_METADATA_V2 = {
    "format-version": 2,
    "table-uuid": "9c12d441-03fe-4693-9a96-a0705ddf69c1",
    "location": "s3://bucket/test/location",
    "last-sequence-number": 34,
    "last-updated-ms": 1602638573590,
    "last-column-id": 3,
    "current-schema-id": 1,
    "schemas": [
        {"type": "struct", "schema-id": 0, "fields": [{"id": 1, "name": "x", "required": True, "type": "long"}]},
        {
            "type": "struct",
            "schema-id": 1,
            "identifier-field-ids": [1, 2],
            "fields": [
                {"id": 1, "name": "x", "required": True, "type": "long"},
                {"id": 2, "name": "y", "required": True, "type": "long", "doc": "comment"},
                {"id": 3, "name": "z", "required": True, "type": "long"},
            ],
        },
    ],
    "default-spec-id": 0,
    "partition-specs": [{"spec-id": 0, "fields": [{"name": "x", "transform": "identity", "source-id": 1, "field-id": 1000}]}],
    "last-partition-id": 1000,
    "default-sort-order-id": 3,
    "sort-orders": [
        {
            "order-id": 3,
            "fields": [
                {"transform": "identity", "source-id": 2, "direction": "asc", "null-order": "nulls-first"},
                {"transform": "bucket[4]", "source-id": 3, "direction": "desc", "null-order": "nulls-last"},
            ],
        }
    ],
    "properties": {"read.split.target.size": "134217728"},
    "current-snapshot-id": 3055729675574597004,
    "snapshots": [
        {
            "snapshot-id": 3051729675574597004,
            "timestamp-ms": 1515100955770,
            "sequence-number": 0,
            "summary": {"operation": "append"},
            "manifest-list": "s3://a/b/1.avro",
        },
        {
            "snapshot-id": 3055729675574597004,
            "parent-snapshot-id": 3051729675574597004,
            "timestamp-ms": 1555100955770,
            "sequence-number": 1,
            "summary": {"operation": "append"},
            "manifest-list": "s3://a/b/2.avro",
            "schema-id": 1,
        },
    ],
    "snapshot-log": [
        {"snapshot-id": 3051729675574597004, "timestamp-ms": 1515100955770},
        {"snapshot-id": 3055729675574597004, "timestamp-ms": 1555100955770},
    ],
    "metadata-log": [{"metadata-file": "s3://bucket/.../v1.json", "timestamp-ms": 1515100}],
    "refs": {"test": {"snapshot-id": 3051729675574597004, "type": "tag", "max-ref-age-ms": 10000000}},
}


metadata_v2 = json.dumps(EXAMPLE_TABLE_METADATA_V2)

for _ in range(10000):
    TableMetadataUtil.parse_raw(metadata_v2)
```

File: python/pyiceberg/io/pyarrow.py [hunk 11]
```diff
@@ -781,7 +781,7 @@ def _task_to_table(
             schema_raw = metadata.get(ICEBERG_SCHEMA)
         # TODO: if field_ids are not present, Name Mapping should be implemented to look them up in the table schema,
         #  see https://github.com/apache/iceberg/issues/7451
-        file_schema = Schema.parse_raw(schema_raw) if schema_raw is not None else pyarrow_to_schema(physical_schema)
+        file_schema = Schema.model_validate_json(schema_raw) if schema_raw is not None else pyarrow_to_schema(physical_schema)
```
Comment (rdblue) [inline python/pyiceberg/io/pyarrow.py:784]: Nit: It uses "validate" instead of "parse"? That seems weird to me.

File: python/pyiceberg/partitioning.py [hunk 12]
```diff
@@ -23,15 +25,21 @@
...
 
 from pyiceberg.schema import Schema
-from pyiceberg.transforms import Transform
+from pyiceberg.transforms import Transform, _deserialize_transform
 from pyiceberg.typedef import IcebergBaseModel
 from pyiceberg.types import NestedField, StructType
 
 INITIAL_PARTITION_SPEC_ID = 0
-_PARTITION_DATA_ID_START: int = 1000
+PARTITION_FIELD_ID_START: int = 1000
```
Comment (rdblue) [inline python/pyiceberg/partitioning.py:42]: Why is this needed?

File: python/pyiceberg/table/__init__.py [hunk 13]
```diff
@@ -365,8 +365,8 @@ class AssertDefaultSortOrderId(TableRequirement):
 
 class CommitTableRequest(IcebergBaseModel):
     identifier: Identifier = Field()
-    requirements: List[TableRequirement] = Field(default_factory=list)
-    updates: List[TableUpdate] = Field(default_factory=list)
+    requirements: List[SerializeAsAny[TableRequirement]] = Field(default_factory=list)
+    updates: List[SerializeAsAny[TableUpdate]] = Field(default_factory=list)
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:369]: Can you help me understand what's happening here?

File: python/pyiceberg/table/metadata.py [hunk 14]
```diff
@@ -28,11 +30,11 @@
 
 from pydantic import Field
 from pydantic import ValidationError as PydanticValidationError
-from pydantic import root_validator
+from pydantic import model_validator
 from typing_extensions import Annotated
 
 from pyiceberg.exceptions import ValidationError
-from pyiceberg.partitioning import PartitionSpec, assign_fresh_partition_spec_ids
+from pyiceberg.partitioning import PARTITION_FIELD_ID_START, PartitionSpec, assign_fresh_partition_spec_ids
```
Comment (rdblue) [inline python/pyiceberg/table/metadata.py:37]: Ah, so this was needed outside of `partitioning` and that's why it is no longer prefixed with an underscore?

File: python/pyiceberg/table/metadata.py (hunk 6 above)
Comment (rdblue) [reply python/pyiceberg/table/metadata.py:83]: I like how we use the same keys now.

File: python/pyiceberg/table/metadata.py [hunk 15]
```diff
@@ -217,7 +232,15 @@ class TableMetadataV1(TableMetadataCommonFields, IcebergBaseModel):
     # because bumping the version should be an explicit operation that is up
     # to the owner of the table.
 
-    @root_validator
+    @model_validator(mode="before")
+    def cleanup_snapshot_id(cls, data: Dict[str, Any]) -> Dict[str, Any]:
+        return cleanup_snapshot_id(data)
```
Comment (rdblue) [inline python/pyiceberg/table/metadata.py:237]: Is this reused? Why have a separate method instead of inlining the logic here?

File: python/pyiceberg/table/snapshots.py [hunk 16]
```diff
@@ -59,34 +58,19 @@ class Summary(IcebergBaseModel):
...
-        return values
-
-    def __init__(
-        self, operation: Optional[Operation] = None, __root__: Optional[Dict[str, Union[str, Operation]]] = None, **data: Any
-    ) -> None:
-        super().__init__(__root__={"operation": operation, **data} if not __root__ else __root__)
-        self._additional_properties = {
-            k: v for k, v in self.__root__.items() if k != OPERATION  # type: ignore # We know that they are all string, and we don't want to check
-        }
+    def __init__(self, operation: Optional[Operation] = None, **data: Any) -> None:
```
Comment (rdblue) [inline python/pyiceberg/table/snapshots.py:64]: While `operation` was not originally written, it should always be there so we don't necessarily need it to be `Optional`.

File: python/pyiceberg/types.py [hunk 17]
```diff
@@ -133,32 +176,34 @@ class FixedType(PrimitiveType):
         False
     """
 
-    __root__: str = Field()
-    _len: int = PrivateAttr()
+    root: int = Field()
```
Comment (rdblue) [inline python/pyiceberg/types.py:179]: Why is the length called `root` now? Can we call it `len` or similar instead? I think it is confusing this way.

File: python/pyiceberg/types.py [hunk 18]
```diff
@@ -167,44 +212,45 @@ class DecimalType(PrimitiveType):
         True
     """
 
-    __root__: str = Field()
+    root: Tuple[int, int]
```
Comment (rdblue) [inline python/pyiceberg/types.py:215]: Similar here, it's odd to keep the type state as "root" unless this is something pydantic requires.

File: python/pyiceberg/types.py [hunk 19]
```diff
@@ -265,7 +311,7 @@ def __str__(self) -> str:
 
     def __getnewargs__(self) -> Tuple[int, str, IcebergType, bool, Optional[str]]:
         """A magic function for pickling the NestedField class."""
-        return (self.field_id, self.name, self.field_type, self.required, self.doc)
+        return self.field_id, self.name, self.field_type, self.required, self.doc
```
Comment (rdblue) [inline python/pyiceberg/types.py:314]: Is this change required?

File: python/pyiceberg/types.py (hunk 18 above)
Comment (Fokko) [reply python/pyiceberg/types.py:215]: This is required. Because this doesn't (de)serializes to dict, it has to map onto the root.

File: python/pyiceberg/types.py [hunk 20]
```diff
@@ -390,25 +446,49 @@ def __init__(
...
-
-        data["key_field"] = NestedField(name="key", field_id=data["key_id"], field_type=data["key_type"], required=True)
-        data["value_field"] = NestedField(
-            name="value", field_id=data["value_id"], field_type=data["value_type"], required=data["value_required"]
-        )
+        data["key-id"] = data["key-id"] if "key-id" in data else key_id
+        data["key"] = data["key"] if "key" in data else key_type
+        data["value-id"] = data["value-id"] if "value-id" in data else value_id
+        data["value"] = data["value"] if "value" in data else value_type
+        data["value_required"] = data["value_required"] if "value_required" in data else value_required
```
Comment (rdblue) [inline python/pyiceberg/types.py:453]: Should this be `value-required`?

File: python/pyiceberg/utils/parsing.py [hunk 21]
```diff
@@ -32,4 +34,4 @@ def match(self, str_repr: str) -> int:
         matches = self.regex.search(str_repr)
         if matches:
             return int(matches.group(1))
-        raise ValueError(f"Could not match {str_repr}, expected format {self.prefix}[22]")
+        raise ValidationError(f"Could not match {str_repr}, expected format {self.prefix}[22]")
```
Comment (rdblue) [inline python/pyiceberg/utils/parsing.py:37]: Should `22` be hard-coded here?

File: python/pyiceberg/io/pyarrow.py (hunk 11 above)
Comment (Fokko) [reply python/pyiceberg/io/pyarrow.py:784]: From the migration guide: https://docs.pydantic.dev/latest/migration/
image

File: python/tests/conftest.py [hunk 22]
```diff
@@ -311,7 +311,7 @@ def all_avro_types() -> Dict[str, Any]:
             ],
         }
     ],
-    "properties": {"read.split.target.size": 134217728},
+    "properties": {"read.split.target.size": "134217728"},
```
Comment (rdblue) [inline python/tests/conftest.py:314]: Good catch.

File: python/pyiceberg/partitioning.py (hunk 12 above)
Comment (Fokko) [reply python/pyiceberg/partitioning.py:42]: When we read a V1 spec, we use this to set the `last-partition-id` on the metadata:
```
        data[LAST_PARTITION_ID] = max(
            [field.get(FIELD_ID) for spec in data[PARTITION_SPECS] for field in spec[FIELDS]], default=PARTITION_FIELD_ID_START
        )
```
This is outside of the package, so I thought it would make sense to make public.

**Comment by rdblue** [comment]
+1 overall. I don't see any blockers once tests are passing. There was one string that looked suspicious since it wasn't translated from underscores to dashes.

File: python/pyiceberg/table/__init__.py (hunk 13 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:369]: I also had to learn this by raising an issue: https://github.com/pydantic/pydantic/issues/6403#issuecomment-1620252270

It turns out that you need to wrap this when you're serializing an object, that might have subclasses. This way it gets correctly serialized.

File: python/pyiceberg/table/metadata.py (hunk 14 above)
Comment (Fokko) [reply python/pyiceberg/table/metadata.py:37]: Exactly, the `_` means that it shouldn't be used outside of the class.

File: python/pyiceberg/transforms.py [hunk 23]
```diff
@@ -106,47 +100,37 @@ def _transform_literal(func: Callable[[L], L], lit: Literal[L]) -> Literal[L]:
     return literal(func(lit.value))
 
 
-class Transform(IcebergBaseModel, ABC, Generic[S, T]):
+def _deserialize_transform(v: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (Fokko) [reply python/pyiceberg/transforms.py:103]: I've renamed this to `parse_transform`

File: python/pyiceberg/types.py (hunk 17 above)
Comment (Fokko) [reply python/pyiceberg/types.py:179]: This is because it serializes to a single value instead of a dict, so we have to map it onto root.

File: python/pyiceberg/types.py (hunk 19 above)
Comment (Fokko) [reply python/pyiceberg/types.py:314]: No, but my IDE was bugging me :)

Comment (Fokko) [reply python/pyiceberg/types.py:314]: Reverted, also for the `{List,Map}Type`

File: python/pyiceberg/table/metadata.py (hunk 6 above)
Comment (Fokko) [reply python/pyiceberg/table/metadata.py:83]: Yes, and it also handles the aliases. For example, for the `NestedField` we both accept `field_id` and `id` as the `id`. But since `id` is a reserved keyword we map `id` to `field_id`. In the `after` validator, we make sure that `id` is converted to `field_id`, which makes it less error-prone :)

File: python/pyiceberg/types.py (hunk 20 above)
Comment (Fokko) [reply python/pyiceberg/types.py:453]: Yes, thank you :)

File: python/pyiceberg/utils/parsing.py (hunk 21 above)
Comment (Fokko) [reply python/pyiceberg/utils/parsing.py:37]: It is an example, it can be arbitrary. The `ParseNumberFromBrackets('ryan')` would accept anything like `ryan[22]`

File: python/pyiceberg/io/pyarrow.py (hunk 11 above)
Comment (Fokko) [reply python/pyiceberg/io/pyarrow.py:784]: I agree, I would also prefer `parse`

**Comment by rdblue** [comment]
Thanks, @Fokko!

**Comment by pdames** [comment]
@Fokko and @rdblue - unfortunately, this has broken our in-progress integration of PyIceberg with Ray on the [Iceberg Branch of DeltaCAT](https://github.com/ray-project/deltacat/tree/iceberg) due to https://github.com/ray-project/ray/issues/37019. Happy to work on both sides to find a path forward, but seems like we'll need to pin to a prior commit of PyIceberg that depends on Pydantic for continued Ray 2.X interoperability in the mean time.

**Comment by blublinsky** [comment]
We have to be very careful here. Pydantic is also used by KFP and Ray. We are currently on Version: 1.10.4. I am afraid, this update will break the compatibility with other components that we are using
